syntax = "proto3";

package iterators;

message TriplePattern {
  string subject = 1;
  string predicate = 2;
  string object = 3;
  string graph = 4;
}

message SolutionMapping {
  map<string, string> bindings = 1;
}

message SavedScanIterator {
  TriplePattern pattern = 1;
  map<string, string> muc = 2;
  map<string, string> mu = 3;
  string last_read = 4;
  string timestamp = 5;
  int64 cardinality = 6;
  int64 cumulative_cardinality = 7;
  int64 pattern_cardinality = 8;
  int64 pattern_produced = 9;
  int64 produced = 10;
  int64 stages = 11;
}

message SavedValuesIterator {
  repeated SolutionMapping values = 1;
  int64 next_value = 2;
  int64 produced = 3;
  map<string, string> muc = 4;
}

message SavedProjectionIterator {
  repeated string values = 1;
  oneof source {
    SavedScanIterator scan_source = 2;
    SavedIndexJoinIterator join_source = 3;
    SavedBagUnionIterator union_source = 4;
    SavedFilterIterator filter_source = 5;
    SavedValuesIterator values_source = 6;
    SavedLeftJoinIterator leftjoin_source = 7;
  }
}

message SavedIndexJoinIterator {
  oneof left {
    SavedScanIterator scan_left = 1;
    SavedProjectionIterator proj_left = 2;
    SavedBagUnionIterator union_left = 3;
    SavedIndexJoinIterator join_left = 4;
    SavedFilterIterator filter_left = 5;
    SavedValuesIterator values_left = 6;
    SavedLeftJoinIterator leftjoin_left = 7;
  }
  oneof right {
    SavedScanIterator scan_right = 8;
    SavedProjectionIterator proj_right = 9;
    SavedBagUnionIterator union_right = 10;
    SavedIndexJoinIterator join_right = 11;
    SavedFilterIterator filter_right = 12;
    SavedValuesIterator values_right = 13;
    SavedLeftJoinIterator leftjoin_right = 14;
  }
  map<string, string> muc = 15;
}

message SavedLeftJoinIterator {
  oneof left {
    SavedScanIterator scan_left = 1;
    SavedProjectionIterator proj_left = 2;
    SavedBagUnionIterator union_left = 3;
    SavedIndexJoinIterator join_left = 4;
    SavedFilterIterator filter_left = 5;
    SavedValuesIterator values_left = 6;
    SavedLeftJoinIterator leftjoin_left = 7;
  }
  oneof right {
    SavedScanIterator scan_right = 8;
    SavedProjectionIterator proj_right = 9;
    SavedBagUnionIterator union_right = 10;
    SavedIndexJoinIterator join_right = 11;
    SavedFilterIterator filter_right = 12;
    SavedValuesIterator values_right = 13;
    SavedLeftJoinIterator leftjoin_right = 14;
  }
  map<string, string> muc = 15;
}

message SavedBagUnionIterator {
  oneof left {
    SavedScanIterator scan_left = 1;
    SavedProjectionIterator proj_left = 2;
    SavedBagUnionIterator union_left = 3;
    SavedIndexJoinIterator join_left = 4;
    SavedFilterIterator filter_left = 5;
    SavedValuesIterator values_left = 6;
    SavedLeftJoinIterator leftjoin_left = 7;
  }
  oneof right {
    SavedScanIterator scan_right = 8;
    SavedProjectionIterator proj_right = 9;
    SavedBagUnionIterator union_right = 10;
    SavedIndexJoinIterator join_right = 11;
    SavedFilterIterator filter_right = 12;
    SavedValuesIterator values_right = 13;
    SavedLeftJoinIterator leftjoin_right = 14;
  }
}

message SavedFilterIterator {
  oneof source {
    SavedScanIterator scan_source = 1;
    SavedProjectionIterator proj_source = 2;
    SavedFilterIterator filter_source = 3;
    SavedIndexJoinIterator join_source = 4;
    SavedValuesIterator values_right = 5;
    SavedLeftJoinIterator leftjoin_source = 6;
  }
  string expression = 7;
  map<string, string> mu = 8;
  int64 consumed = 9;
  int64 produced = 10;
}

message SavedInsertData {
  map<string, uint64> nb_inserted = 1;
}

message SavedDeleteData {
  map<string, uint64> nb_inserted = 1;
}

message RootTree {
  oneof source {
    SavedScanIterator scan_source = 1;
    SavedProjectionIterator proj_source = 2;
    SavedIndexJoinIterator join_source = 3;
    SavedBagUnionIterator union_source = 4;
    SavedFilterIterator filter_source = 5;
    SavedInsertData insert_source = 6;
    SavedDeleteData delete_source = 7;
    SavedValuesIterator values_right = 8;
    SavedLeftJoinIterator leftjoin_source = 9;
  }
}
